modules/qoa/audio_stream_qoa.cpp

```diff
@@ -0,0 +1,281 @@
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#define QOA_IMPLEMENTATION
+#define QOA_NO_STDIO
+
+#include "audio_stream_qoa.h"
+
+#include "core/io/file_access.h"
+
+int AudioStreamPlaybackQOA::_mix_internal(AudioFrame *p_buffer, int p_frames) {
+	if (!active) {
+		return 0;
+	}
+
+	int todo = p_frames;
+	int frames_mixed_this_step = p_frames;
+
+	uint32_t begin_limit = (qoa_stream->loop_mode != AudioStreamQOA::LOOP_DISABLED) ? qoa_stream->loop_begin : 0;
+	uint32_t end_limit = (qoa_stream->loop_mode != AudioStreamQOA::LOOP_DISABLED) ? qoa_stream->loop_end : qoad->samples;
+
+	while (todo && active) {
+		int src_index = decoded_offset * qoa_stream->channels;
+
+		if (decoded_len <= decoded_offset) {
+			// Decode the next QOA frame
+			data_offset += frame_data_len;
+			qoa_decode_frame(qoa_stream->data.ptr() + data_offset, frame_data_len, qoad, decoded, &decoded_len);
+			decoded_offset = 0;
+			src_index = 0;
+		}
+
+		p_buffer[p_frames - todo][0] = decoded[qoa_stream->channels == 1 ? src_index : src_index++];
+		p_buffer[p_frames - todo][1] = decoded[src_index++];
+		p_buffer[p_frames - todo] /= 32767.0f;
+
+		// Discard channels other than L and R
+		if (qoa_stream->channels > 2) {
+			src_index += qoa_stream->channels - 2;
+		}
```

Comment:
I'm assuming that `qoa_stream->channels` stays the same during the stream's lifetime. So how about precalculating a new member variable `skipped_channels` (maybe in `instantiate_playback()`). Something like this:

```c++
skipped_channels = MAX(0, channels - 2);
```

Now the `if` could be removed here altogether:

```c++
src_index += qoa_stream.skipped_channels;
```

I'm assuming that an addition would be faster than a conditional, but as always when optimizing, some actual testing should be done.

Reply:
Done.

Reply:
Upon further inspection, I realized that since `src_index` is re-generated every iteration, discarding channels is not necessary at all. It's leftover of a previous buggy implementation that did require it.